Draft: Updates to the reset/reboot/remove node procedures

paigecalvert opened the pull request:
Story: https://app.shortcut.com/replicated/story/59562/kurl-node-removal-updates

Latest preview: https://634d8cc5bb9f0600567d15d7--kurlsh-staging.netlify.app/docs/install-with-kurl/managing-nodes/

laverya commented on src/markdown-pages/install-with-kurl/managing-nodes.md:
```diff
@@ -0,0 +1,247 @@
+---
+path: "/docs/install-with-kurl/managing-nodes"
+date: "2019-12-19"
```
```suggestion
date: "2022-10-12"
```

laverya commented on src/markdown-pages/install-with-kurl/managing-nodes.md:
```diff
@@ -0,0 +1,247 @@
+
+* In the kURL specification, set `isBlockStorageEnabled` to `true`. This is the default for Rook Ceph v1.4 and later.
+
+* Ensure that you can access the ceph CLI from a Pod that can communicate with the Ceph Storage Cluster. To access the ceph CLI, you can do one of the following:      
+
+  * (Recommended) Use the `rook-ceph-tools` Pod to access the ceph CLI. Use the same version of the Rook toolbox as the version of Rook Ceph that is installed in the cluster.
+
+     For information about how to create a `rook-ceph-tools` Pod for Rook Ceph v1.5, see [Rook Toolbox](https://rook.io/docs/rook/v1.5/ceph-toolbox.html) in the Rook Ceph documentation.
```
this indentation is not actually respected.

Also, I think we ship rook-ceph-tools by default on all 1.4+ installs.

laverya commented on src/markdown-pages/install-with-kurl/managing-nodes.md:
````diff
@@ -0,0 +1,247 @@
+    Open an interactive shell in the `rook-ceph-tools` or `rook-ceph-operator` Pod:
+
+    ```
+    kubectl exec -it -n rook-ceph rook-ceph-tools-54ff78f9b6-gqsfm -- /bin/bash
````
```suggestion
    kubectl exec -it -n rook-ceph deployment/rook-ceph-tools -- bash
```

laverya commented on src/markdown-pages/install-with-kurl/managing-nodes.md:
````diff
@@ -0,0 +1,247 @@
+    **Example**:
+
+    ```
+    kubectl exec -it -n rook-ceph rook-ceph-tools-54ff78f9b6-gqsfm -- ceph status
````
```suggestion
    kubectl exec -it -n rook-ceph deployment/rook-ceph-tools -- ceph status
```

laverya commented on src/markdown-pages/install-with-kurl/managing-nodes.md:
```diff
@@ -0,0 +1,247 @@
+
+### (Recommended) Manually Rebalance Ceph and Remove a Node
+
+This procedure ensures that the data held in Rook Ceph is safely replicated to a new node before you remove a node. Rebalancing your data is critical for preventing against data loss that can occur when removing a node if the data stored in Ceph has not been properly replicated.
```
```suggestion
This procedure ensures that the data held in Rook Ceph is safely replicated to a new node before you remove a node.
Rebalancing your data is critical for preventing data loss that can occur when removing a node if the data stored in Ceph has not been properly replicated.
```

jonquil2002 replied:
I like Andrew's suggestion.

laverya commented on src/markdown-pages/install-with-kurl/managing-nodes.md:
```diff
@@ -0,0 +1,247 @@
+
+This procedure ensures that the data held in Rook Ceph is safely replicated to a new node before you remove a node. Rebalancing your data is critical for preventing against data loss that can occur when removing a node if the data stored in Ceph has not been properly replicated.
+
+To manually remove a node, you first use the ceph CLI to weigh the Ceph OSD to `0` on the node that you want to remove and wait for Ceph to rebalance the data across OSDs. Then, you can remove the OSD from the node, and finally remove the node.
```
```suggestion
To manually remove a node, you first use the Ceph CLI to reweight the Ceph OSD to `0` on the node that you want to remove and wait for Ceph to rebalance the data across OSDs.
Then, you can remove the OSD from the node, and finally remove the node.
```

jonquil2002 replied:
You could also make these an ordered sublist, though it's hard to tell without the preview.

laverya commented on src/markdown-pages/install-with-kurl/managing-nodes.md:
```diff
@@ -0,0 +1,247 @@
+
+1. Complete the [Rook Ceph Cluster Prerequisites](#rook-ceph-cluster-prerequisites) above.
+
+1. Add the same number of new nodes to the cluster that you intended to remove. For example, if you intend to remove a total of two nodes, add two new nodes.
```
```suggestion
1. Add the same number of new nodes to the cluster that you intended to remove.
For example, if you intend to remove a total of two nodes, add two new nodes.
```

paigecalvert replied:
@jonquil2002 for context, I think this change has to do with Andrew's/the kURL team's preference to have one line of markdown per sentence (for ease of viewing changes in a git diff). Though correct me if I'm wrong about the intent Andrew

laverya commented on src/markdown-pages/install-with-kurl/managing-nodes.md:
````diff
@@ -0,0 +1,247 @@
+   ```
+   ceph osd reweight OSD_NAME 0
+   ```
+   Replace `OSD_NAME` with the name of the Ceph OSD on the node that you intend to remove.
````
It's not a name, it's a number
The full command looks like `ceph osd reweight 1 0` for OSD 1 etc

laverya commented on src/markdown-pages/install-with-kurl/managing-nodes.md:
````diff
@@ -0,0 +1,247 @@
+   ```
+   Replace `OSD_NAME` with the name of the Ceph OSD on the node that you intend to remove.
+
+   Ceph rebalances the placement groups off the OSD that you specify in the `ceph osd reweight` command. To view progress, run `watch ceph status`.
````
```suggestion
   Ceph rebalances the placement groups off the OSD that you specify in the `ceph osd reweight` command.
   To view progress, run `ceph status`.
```
IMO telling people to run `watch` is a little weird

paigecalvert replied:
Works for me. For context, I pulled this suggestion from https://community.replicated.com/t/how-to-safely-resize-a-rook-ceph-cluster/991

laverya commented on src/markdown-pages/install-with-kurl/managing-nodes.md:
````diff
@@ -0,0 +1,247 @@
+   kubectl scale deployment -n rook-ceph rook-ceph-osd-1 --replicas 0
+   ```   
+
+1. Remove the node.
````
If we're going through all this trouble to scale things down etc, we really need to make people run the ok-to-remove command

laverya replied:
It looks like this
```
[rook@rook-ceph-tools-6fb84b545-js4rg /]$ ceph osd safe-to-destroy osd.1
OSD(s) 1 are safe to destroy without reducing data durability.
```

paigecalvert replied:
Thanks, I'll add a step for this

laverya commented on src/markdown-pages/install-with-kurl/managing-nodes.md:
```diff
@@ -0,0 +1,247 @@
+
+1. Remove the node.
+
+   After the node is removed, Ceph replicates its data to OSDs on remaining nodes. If a Ceph OSD is scheduled on the node that you removed, verify that Ceph is in a healthy state.
```
```suggestion
   After the node is removed, Ceph replicates its data to OSDs on remaining nodes.
   If a Ceph OSD was scheduled on the node that you removed, verify that Ceph is in a healthy state.
```

laverya commented on src/markdown-pages/install-with-kurl/managing-nodes.md:
```diff
@@ -0,0 +1,247 @@
+
+   After the node is removed, Ceph replicates its data to OSDs on remaining nodes. If a Ceph OSD is scheduled on the node that you removed, verify that Ceph is in a healthy state.
+
+1. (Optional) After the node is removed, you can run the kURL reset script to remove kURL and Kubernetes assets from the node to prep it to re-join the cluster at a later time. Or, delete the VM and provision a new VM.
```
```suggestion
1. (Optional) After the node is removed, you can run the kURL reset script to remove kURL and Kubernetes assets from the node to prep it to re-join the cluster at a later time. Or, delete the VM and provision a new VM.
```

You can't rejoin a node to a cluster after running the reset script IIRC

It's just not a good idea

paigecalvert replied:
We can probably just delete this whole step, then (unless you object)

jonquil2002 commented on src/markdown-pages/install-with-kurl/managing-nodes.md:
```diff
@@ -0,0 +1,247 @@
+   * You must have at least one primary node.
+   * If you scale the etcd cluster to three primary nodes, you must then maintain a minimum of three primary nodes to maintain quorum.
+* **Rook Ceph cluster health**: When you scale a Ceph Storage Cluster to three or more Ceph Object Storage Daemons (OSDs), such as when you add additional manager or worker nodes to the cluster, the Ceph Storage Cluster can no longer have fewer than three OSDs. If you reduce the number of OSDs to less than three in this case, then the Ceph Storage Cluster loses quorum.
+* **Add a node before removing a node**: To remove and replace a node, it is recommended that you add a new node first before removing the node. For example, to remove one node in a three-node cluster, first add a new node to scale the cluster to four nodes. Then, remove the desired node to scale the cluster back down to three nodes.
```
This paragraph uses "first" twice. 

s/ To remove and replace a node, it is recommended that you add a new node first before removing the node. / To remove and replace a node, it is recommended that you add a new node before removing the node.

jonquil2002 commented on src/markdown-pages/install-with-kurl/managing-nodes.md:
```diff
@@ -0,0 +1,247 @@
+
+     You can either open an interactive shell to run multiple ceph CLI commands in a row, or you can precede each ceph CLI command that you run with `kubectl exec`. The following shows an example of each method:
+
+ * (Optional) Open an interactive shell in the `rook-ceph-tools` or `rook-ceph-operator` Pod to run multiple ceph CLI commands in a row.
```
It's hard to tell without a preview, but I think it might be easier to read if you move this as an intro after "Example:"

And start this option with "**Example: Using an Interactive Shell** 

And then do something similar with the other method (non-interactive shell).

paigecalvert replied:
Ah good call, yeah, this section is unnecessarily complex in terms of these little headings. Reworking

jonquil2002 commented on src/markdown-pages/install-with-kurl/managing-nodes.md:
```diff
@@ -0,0 +1,247 @@
+
+   Ceph rebalances the placement groups off the OSD that you specify in the `ceph osd reweight` command. To view progress, run `watch ceph status`.
+
+1. After the `ceph osd reweight` command completes, verify that Ceph is in a healthy state:
```
s/ verify that Ceph is in a healthy state: / run the following command to verify that Ceph is in a healthy state:

There are a few steps in this section that are inconsistent with our usual "run:" or "run the following command to" style.

jonquil2002 commented on src/markdown-pages/install-with-kurl/managing-nodes.md:
````diff
@@ -0,0 +1,247 @@
+       kubectl -n rook-ceph exec deploy/rook-ceph-tools -- ceph status
+       ```
+
+    * **Data replication**: A common Ceph configuration is three data replicas across three Ceph OSDs. It is possible for Ceph to report a healthy status without data being replicated properly across all OSDs. For example, in a single-node cluster, there are not multiple machines where Ceph can replicate data. In this case, even if Ceph reports healthy, removing a node will result in data loss be ...
````
s/ will result / results

jonquil2002 commented on src/markdown-pages/install-with-kurl/managing-nodes.md:
```diff
@@ -0,0 +1,247 @@
+
+    * **Data replication**: A common Ceph configuration is three data replicas across three Ceph OSDs. It is possible for Ceph to report a healthy status without data being replicated properly across all OSDs. For example, in a single-node cluster, there are not multiple machines where Ceph can replicate data. In this case, even if Ceph reports healthy, removing a node will result in data loss be ...
+
+    If you are not certain that Ceph data replication was configured and completed properly, or if Ceph is not in a healthy state, it is recommended that you instead follow the procedure to first rebalance the data off the node that you intend to remove to avoid data loss. See [(Recommended) Manually Rebalance Ceph and Remove a Node](#ecommended-manually-rebalance-ceph-and-remove-a-node) above.
```
s/ it is recommended that you instead follow the procedure to first rebalance the data off the node that you intend to remove to avoid data loss / it is recommended that you first rebalance the data off the node that you intend to remove to avoid data loss

paigecalvert commented:
@laverya thanks again for the review. I incorporated your feedback, and also updated the markdown header content, added this new topic to the TOC, and updated the content to be one-line-per-sentence. 

Will you give this a last pass/approval when you have a moment? We should be able to merge after you sign off, since Chris already gave approval on the content from the Support team.

paigecalvert commented:
Closing this PR in favor of https://github.com/replicatedhq/kurl.sh/pull/888
